File: realtime/ios/webrtc/Sources/State/ConversationViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class ConversationViewModel {
    private(set) var transcript: [TranscriptItem] = []
    private(set) var state: SessionState = .idle
    var isMicMuted = false {
        didSet { session?.setMicEnabled(!isMicMuted) }
    }

    private let settings: SettingsStore
    private var session: RealtimeSession?
    private var streamingAgentItemID: UUID?
    private var streamingUserItemID: UUID?

    init(settings: SettingsStore) {
        self.settings = settings
    }

    var isConnected: Bool { state == .connected }
    var isBusy: Bool { state == .connecting }

    var errorMessage: String? {
        if case .failed(let message) = state { return message }
        return nil
    }

    func connect() async {
        transcript = []
        streamingAgentItemID = nil
        streamingUserItemID = nil
        isMicMuted = false

        let session = RealtimeSession(
            authProvider: settings.makeAuthProvider(),
            config: settings.makeSessionConfig()
        )
        session.onStateChange = { [weak self] state in self?.state = state }
        session.onEvent = { [weak self] event in self?.handle(event) }
        self.session = session
        await session.connect()
    }

    func disconnect() {
        session?.disconnect()
        session = nil
        streamingAgentItemID = nil
        streamingUserItemID = nil
    }

    func handle(_ event: ServerEvent) {
        switch event {
        case .outputTextDelta(let delta):
            appendAgentDelta(delta)
        case .transcriptDone(let text):
            finalizeAgentItem(replacingWith: text)
        case .inputTranscriptionDelta(let delta):
            appendUserDelta(delta)
        case .inputTranscriptionCompleted(let transcript):
            finalizeUserItem(with: transcript)
        case .speechStarted:
            dropStreamingAgentItem()
        case .responseDone:
            finalizeAgentItem(replacingWith: nil)
        case .error(let message):
            state = .failed(message)
        case .outputItemAdded, .unknown,
             .backchannelAudioDelta, .backchannelAudioDone, .backchannelSkipped:
            // Back-channel audio is played in the realtime layer; phrase is telemetry, not transcript.
            break
        }
    }

    private func appendAgentDelta(_ delta: String) {
        guard !delta.isEmpty else { return }
        if let id = streamingAgentItemID, let index = transcript.firstIndex(where: { $0.id == id }) {
            transcript[index].text += delta
        } else {
            var item = TranscriptItem(role: .agent, text: delta)
            item.isStreaming = true
            streamingAgentItemID = item.id
            transcript.append(item)
        }
    }

    private func finalizeAgentItem(replacingWith text: String?) {
        guard let id = streamingAgentItemID,
              let index = transcript.firstIndex(where: { $0.id == id }) else {
            streamingAgentItemID = nil
            return
        }
        if let text, !text.isEmpty {
            transcript[index].text = text
        }
        transcript[index].isStreaming = false
        streamingAgentItemID = nil
    }

    private func appendUserDelta(_ delta: String) {
        guard !delta.isEmpty else { return }
        if let id = streamingUserItemID, let index = transcript.firstIndex(where: { $0.id == id }) {
            transcript[index].text = Self.reconcileTranscript(existing: transcript[index].text, delta: delta)
        } else {
            var item = TranscriptItem(role: .user, text: delta)
            item.isStreaming = true
            streamingUserItemID = item.id
            transcript.append(item)
        }
    }

    /// Some STT providers (e.g. Soniox) emit each partial as the FULL text-so-far rather
    /// than an incremental chunk, so blindly appending duplicates the transcript. The
    /// realtime contract intends `delta` to be incremental, but until the server normalizes
    /// it, tolerate both shapes. The final `completed` transcript is authoritative regardless.
    static func reconcileTranscript(existing: String, delta: String) -> String {
        if existing.isEmpty { return delta }
        if delta == existing { return existing }        // cumulative re-send of same text
        if delta.hasPrefix(existing) { return delta }   // cumulative growth → replace
        if existing.hasPrefix(delta) { return existing } // stale shorter snapshot → keep
        return existing + delta                          // genuine incremental chunk → append
    }

    private func finalizeUserItem(with transcript: String) {
        let trimmed = transcript.trimmingCharacters(in: .whitespacesAndNewlines)
        if let id = streamingUserItemID, let index = self.transcript.firstIndex(where: { $0.id == id }) {
            if !trimmed.isEmpty {
                self.transcript[index].text = trimmed
                self.transcript[index].isStreaming = false
            } else if self.transcript[index].text.isEmpty {
                // Nothing streamed and no final text: drop the empty bubble.
                self.transcript.remove(at: index)
            } else {
                // Empty final but real partials arrived: keep them, just stop streaming.
                self.transcript[index].isStreaming = false
            }
            streamingUserItemID = nil
        } else if !trimmed.isEmpty {
            // No partials arrived — append the final transcript directly.
            self.transcript.append(TranscriptItem(role: .user, text: trimmed))
        }
    }

    private func dropStreamingAgentItem() {
        if let id = streamingAgentItemID {
            transcript.removeAll { $0.id == id }
            streamingAgentItemID = nil
        }
    }
}

File: realtime/android/webrtc/app/src/main/java/ai/inworld/voiceagent/state/ConversationViewModel.kt
package ai.inworld.voiceagent.state

import ai.inworld.voiceagent.realtime.RealtimeSessionApi
import ai.inworld.voiceagent.realtime.events.ServerEvent
import ai.inworld.voiceagent.storage.Settings
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ConversationUiState(
    val transcript: List<TranscriptItem> = emptyList(),
    val sessionState: SessionState = SessionState.Idle,
    val isMicMuted: Boolean = false,
) {
    val isConnected: Boolean get() = sessionState == SessionState.Connected
    val isBusy: Boolean get() = sessionState == SessionState.Connecting
    val errorMessage: String? get() = (sessionState as? SessionState.Failed)?.message
}

class ConversationViewModel(
    private val sessionFactory: (Settings, kotlinx.coroutines.CoroutineScope) -> RealtimeSessionApi,
) : ViewModel() {
    private val _uiState = MutableStateFlow(ConversationUiState())
    val uiState: StateFlow<ConversationUiState> = _uiState

    private var session: RealtimeSessionApi? = null
    private var collectJobs = mutableListOf<Job>()
    private var streamingAgentItemId: String? = null
    private var streamingUserItemId: String? = null

    fun connect(settings: Settings) {
        disconnect() // tear down any previous session and collectors before starting anew
        _uiState.update { it.copy(transcript = emptyList(), isMicMuted = false) }
        streamingAgentItemId = null
        streamingUserItemId = null

        val session = sessionFactory(settings, viewModelScope)
        this.session = session
        collectJobs += session.state
            .onEach { state -> _uiState.update { it.copy(sessionState = state) } }
            .launchIn(viewModelScope)
        collectJobs += session.events
            .onEach(::handle)
            .launchIn(viewModelScope)
        viewModelScope.launch { session.connect() }
    }

    fun disconnect() {
        session?.disconnect()
        session = null
        collectJobs.forEach { it.cancel() }
        collectJobs.clear()
        streamingAgentItemId = null
        streamingUserItemId = null
    }

    fun setMicMuted(muted: Boolean) {
        _uiState.update { it.copy(isMicMuted = muted) }
        session?.setMicEnabled(!muted)
    }

    override fun onCleared() {
        disconnect()
    }

    fun handle(event: ServerEvent) {
        when (event) {
            is ServerEvent.OutputTextDelta -> appendAgentDelta(event.delta)
            is ServerEvent.TranscriptDone -> finalizeAgentItem(replacingWith = event.transcript)
            is ServerEvent.InputTranscriptionDelta -> appendUserDelta(event.delta)
            is ServerEvent.InputTranscriptionCompleted -> finalizeUserItem(event.transcript)
            is ServerEvent.SpeechStarted -> dropStreamingAgentItem()
            is ServerEvent.ResponseDone -> finalizeAgentItem(replacingWith = null)
            is ServerEvent.Error -> _uiState.update { it.copy(sessionState = SessionState.Failed(event.message)) }
            // Back-channel audio is played in the realtime layer; phrase is telemetry, not transcript.
            is ServerEvent.OutputItemAdded, is ServerEvent.Unknown,
            is ServerEvent.BackchannelAudioDelta, is ServerEvent.BackchannelAudioDone,
            is ServerEvent.BackchannelSkipped,
            -> Unit
        }
    }

    val transcript: List<TranscriptItem> get() = _uiState.value.transcript

    private fun updateTranscript(transform: (List<TranscriptItem>) -> List<TranscriptItem>) {
        _uiState.update { it.copy(transcript = transform(it.transcript)) }
    }

    private fun appendAgentDelta(delta: String) {
        if (delta.isEmpty()) return
        val id = streamingAgentItemId
        val index = id?.let { transcript.indexOfFirst { item -> item.id == it } } ?: -1
        if (index >= 0) {
            updateTranscript { list ->
                list.toMutableList().apply { this[index] = this[index].copy(text = this[index].text + delta) }
            }
        } else {
            val item = TranscriptItem(role = Role.Agent, text = delta, isStreaming = true)
            streamingAgentItemId = item.id
            updateTranscript { it + item }
        }
    }

    private fun finalizeAgentItem(replacingWith: String?) {
        val id = streamingAgentItemId ?: return
        streamingAgentItemId = null
        val index = transcript.indexOfFirst { it.id == id }
        if (index < 0) return
        updateTranscript { list ->
            list.toMutableList().apply {
                val text = replacingWith?.takeIf { it.isNotEmpty() } ?: this[index].text
                this[index] = this[index].copy(text = text, isStreaming = false)
            }
        }
    }

    private fun appendUserDelta(delta: String) {
        if (delta.isEmpty()) return
        val id = streamingUserItemId
        val index = id?.let { transcript.indexOfFirst { item -> item.id == it } } ?: -1
        if (index >= 0) {
            updateTranscript { list ->
                list.toMutableList().apply {
                    this[index] = this[index].copy(text = reconcileTranscript(this[index].text, delta))
                }
            }
        } else {
            val item = TranscriptItem(role = Role.User, text = delta, isStreaming = true)
            streamingUserItemId = item.id
            updateTranscript { it + item }
        }
    }

    private fun finalizeUserItem(finalTranscript: String) {
        val trimmed = finalTranscript.trim()
        val id = streamingUserItemId
        val index = id?.let { transcript.indexOfFirst { item -> item.id == it } } ?: -1
        if (index >= 0) {
            updateTranscript { list ->
                list.toMutableList().apply {
                    when {
                        trimmed.isNotEmpty() -> this[index] = this[index].copy(text = trimmed, isStreaming = false)
                        this[index].text.isEmpty() -> removeAt(index) // nothing streamed, no final: drop
                        else -> this[index] = this[index].copy(isStreaming = false) // keep partials
                    }
                }
            }
            streamingUserItemId = null
        } else if (trimmed.isNotEmpty()) {
            // No partials arrived — append the final transcript directly.
            updateTranscript { it + TranscriptItem(role = Role.User, text = trimmed) }
        }
    }

    private fun dropStreamingAgentItem() {
        val id = streamingAgentItemId ?: return
        streamingAgentItemId = null
        updateTranscript { list -> list.filterNot { it.id == id } }
    }

    companion object {
        /** Some STT providers (e.g. Soniox) emit each partial as the FULL text-so-far rather
         *  than an incremental chunk, so blindly appending duplicates the transcript. Tolerate
         *  both shapes; the final `completed` transcript is authoritative regardless. */
        fun reconcileTranscript(existing: String, delta: String): String = when {
            existing.isEmpty() -> delta
            delta == existing -> existing // cumulative re-send of same text
            delta.startsWith(existing) -> delta // cumulative growth → replace
            existing.startsWith(delta) -> existing // stale shorter snapshot → keep
            else -> existing + delta // genuine incremental chunk → append
        }
    }
}
